### PR TITLE
feat(1314): Add src url option

### PR DESCRIPTION
**Pull request by s-yoshika**
## Context
With `src-url` option, we can also build the source code from the remote scm.

## Objective
Add `src-url` option.

Please comment following specifications for `src-url` option
* The URL of `src-url` can be passed with 2 format like below.
  * `https://github.com/<org>/<repo>.git#<branch>`
  * `git@github.com/<org>/<repo>.git#<branch>`
* The source codes are stored under the directory, `~/.sdlocal/repo/<randomId>`,  and cleaned up when the build is finished

## References
https://github.com/screwdriver-cd/screwdriver/issues/1314
https://github.com/screwdriver-cd/screwdriver/blob/master/design/sd-local.md#options

## License

I confirm that this contribution is made under the terms of the license found in the root directory of this repository's source tree and that I have the authority necessary to make this contribution on behalf of its copyright owner.

### CODE REVIEW

File: cmd/build.go [hunk 1]
```diff
@@ -76,8 +104,9 @@ func newBuildCmd() *cobra.Command {
 			}
 			go logger.Run()
 
-			launch := launchNew(job, config, jobName, api.JWT(), artifactsPath)
+			launch := launchNew(job, config, jobName, api.JWT(), artifactsPath, srcPath)
 
+			fmt.Println("Prepare to start build...")
```
Comment (kkisic) [inline cmd/build.go]: Is it better to use `logrus.Info()`?

File: go.mod [hunk 2]
```diff
@@ -9,6 +9,8 @@ require (
 	github.com/sirupsen/logrus v1.4.2
 	github.com/spf13/cobra v0.0.5
 	github.com/stretchr/testify v1.4.0
+	golang.org/x/lint v0.0.0-20200130185559-910be7a94367 // indirect
+	golang.org/x/tools v0.0.0-20200301222351-066e0c02454c // indirect
```
Comment (kkisic) [inline go.mod]: Do we need these?
I think using `go mod tidy` make these unnecessary.

File: cmd/build.go [hunk 3]
```diff
@@ -38,24 +43,47 @@ func newBuildCmd() *cobra.Command {
...
-			config, err := configNew(filepath.Join(homedir, ".sdlocal", "config"))
+			sdlocalDir := filepath.Join(homedir, ".sdlocal")
+			cwd, err := os.Getwd()
 			if err != nil {
 				logrus.Fatal(err)
 			}
+			srcPath := cwd
+
+			if srcURL != "" {
+				fmt.Printf("Pulling the source code from %s...\n", srcURL)
```
Comment (kkisic) [inline cmd/build.go]: Is it better to use logrus.Info()?

File: scm/scm_test.go [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+			baseDir:   baseDir,
+			remoteURL: "https://github.com/screwdriver-cd/sd-local.git",
+			branch:    "test",
+		}
+
+		s, err := New(baseDir, srcURL)
+		defer os.RemoveAll(s.LocalPath())
+
+		expected.localPath = s.LocalPath()
+		assert.Equal(t, expected, s)
```
Comment (kkisic) [inline scm/scm_test.go]: I think it's better to assert only `s.baseDir`, `s.remoteURL` and `s.branch` than to do this

File: scm/scm_test.go [hunk 5]
```diff
@@ -0,0 +1,232 @@
...
+			baseDir:   baseDir,
+			remoteURL: "git@github.com:screwdriver-cd/sd-local.git",
+			branch:    "branch#test",
+		}
+
+		s, err := New(baseDir, srcURL)
+		defer os.RemoveAll(filepath.Join(baseDir, "repo"))
+
+		expected.localPath = s.LocalPath()
+		assert.Equal(t, expected, s)
```
Comment (kkisic) [inline scm/scm_test.go]: Same as above.

File: scm/scm_test.go [hunk 6]
```diff
@@ -0,0 +1,232 @@
...
+	})
+
+	t.Run("failure with making directory", func(t *testing.T) {
+		osMkdirAll = func(path string, per os.FileMode) error { return fmt.Errorf("test") }
+
+		baseDir := os.TempDir()
+		srcURL := "https://github.com/screwdriver-cd/sd-local.git#test"
+
+		scm, err := New(baseDir, srcURL)
+		defer os.RemoveAll(filepath.Join(baseDir, "repo"))
```
Comment (kkisic) [inline scm/scm_test.go]:
```suggestion
		defer os.RemoveAll(s.LocalPath())
```
Forgot to fix?

File: scm/scm_test.go [hunk 7]
```diff
@@ -0,0 +1,232 @@
...
+
+		assert.Nil(t, scm)
+		msg := err.Error()
+		assert.Equal(t, 0, strings.Index(msg, "failed to make local source directory: "), fmt.Sprintf("expected error is `failed to make local source directory: ...`, actual: `%v`", msg))
+
+		_, err = os.Stat(filepath.Join(baseDir, "repo", "github.com", "screwdriver-cd", "sd-local"))
```
Comment (kkisic) [inline scm/scm_test.go]:
```suggestion
		_, err = os.Stat(s.LocalPath())
```
Same as above.

File: scm/scm_test.go [hunk 8]
```diff
@@ -0,0 +1,232 @@
...
+	})
+
+	t.Run("failure with invalid url", func(t *testing.T) {
+		osMkdirAll = func(path string, per os.FileMode) error { return fmt.Errorf("test") }
+
+		baseDir := os.TempDir()
+		srcURL := "https://github.com/screwdriver-cd"
+
+		scm, err := New(baseDir, srcURL)
+		defer os.RemoveAll(filepath.Join(baseDir, "repo"))
```
Comment (kkisic) [inline scm/scm_test.go]:
```suggestion
		defer os.RemoveAll(s.LocalPath())
```
Same as above.

File: scm/scm_test.go [hunk 9]
```diff
@@ -0,0 +1,232 @@
...
+
+		assert.Nil(t, scm)
+		assert.Equal(t, err.Error(), "failed to fetch source code with invalid URL: https://github.com/screwdriver-cd")
+
+		_, err = os.Stat(filepath.Join(baseDir, "repo", "github.com", "screwdriver-cd", "sd-local"))
```
Comment (kkisic) [inline scm/scm_test.go]:
```suggestion
		_, err = os.Stat(s.LocalPath())
```
Same as above.

File: cmd/build.go (hunk 3 above)
Comment (s-yoshika) [reply cmd/build.go]: I'll fix it

File: cmd/build.go (hunk 1 above)
Comment (s-yoshika) [reply cmd/build.go]: I'll fix it

File: go.mod (hunk 2 above)
Comment (s-yoshika) [reply go.mod]: Thanks, I tryed it and these packages doesn't need.

File: scm/scm_test.go (hunk 4 above)
Comment (s-yoshika) [reply scm/scm_test.go]: I'll fix it.

**Comment by kkokufud** [comment]
[ask]
It works `https:`, but `git@` did not work.
Because exec user was `root`, so ssh key for git is changed.
Do we need a discussion about it?

File: scm/scm_test.go (hunk 6 above)
Comment (s-yoshika) [reply scm/scm_test.go]: thanks

File: scm/scm_test.go (hunk 8 above)
Comment (s-yoshika) [reply scm/scm_test.go]: :bow:

File: scm/scm_test.go (hunk 9 above)
Comment (s-yoshika) [reply scm/scm_test.go]: :bow:

**Comment by s-yoshika** [comment]
> [ask]
It works https:, but git@ did not work.
Because exec user was root, so ssh key for git is changed.
Do we need a discussion about it?

It can execute if you pass the ssh-key in somehow like below, I think it's responsible for users to manage ssh-key. And it may be enough with `https` schema just to build, because the source code are deleted if the build is finished.

```
$ sudo SSH_AUTH_SOCK=/tmp/ssh-xxxx/agent.xxxx ./sd-local build --src-url git@github.com:s-yoshika/sdcd-test01.git main
INFO[0000] Pulling the source code from git@github.com:s-yoshika/sdcd-test01.git...
INFO[0000] Prepare to start build...
```

**Comment by sakka2** [comment]
> It can execute if you pass the ssh-key in somehow like below, I think it's responsible for users to manage ssh-key. And it may be enough with https schema just to build, because the source code are deleted if the build is finished.

OK, I see.

**Pull request merged**